source/extensions/clusters/aggregate/cluster_util.cc

```diff
@@ -0,0 +1,61 @@
+#include "extensions/clusters/aggregate/cluster_util.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+std::pair<Upstream::PrioritySetImpl,
+          std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>
+ClusterUtil::linearizePrioritySet(Upstream::ClusterManager& cluster_manager,
+                                  const std::vector<std::string>& clusters) {
+  Upstream::PrioritySetImpl priority_set;
+  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>> priority_to_cluster;
+  int next_priority = 0;
+
+  // Linearize the priority set. e.g. for clusters [C_0, C_1, C_2] referred in aggregate cluster
+  //    C_0 [P_0, P_1, P_2]
+  //    C_1 [P_0, P_1]
+  //    C_2 [P_0, P_1, P_2, P_3]
+  // The linearization result is:
+  //    [C_0.P_0, C_0.P_1, C_0.P_2, C_1.P_0, C_1.P_1, C_2.P_0, C_2.P_1, C_2.P_2, C_2.P_3]
+  // and the traffic will be distributed among these priorities.
+  for (const auto& cluster : clusters) {
+    auto tlc = cluster_manager.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority = 0;
```

Comment:
can you think of some more descriptive names for this + next_priority? looking at the code it's not easy to grok which one is the priority of the current cluster been looked at and the overall linearized priority